felix/wireguard/bootstrap.go

```diff
@@ -17,110 +17,386 @@ package wireguard
+	configParams *config.Config,
+	getNetlinkHandle func() (netlinkshim.Interface, error),
+	getWireguardHandle func() (netlinkshim.Wireguard, error),
+	calicoClient clientv3.Interface,
+) (set.Set, error) {
 	wgDeviceName := configParams.WireguardInterfaceName
 	nodeName := configParams.FelixHostname
 
-	_, dbgBootstrap := os.LookupEnv("FELIX_DBG_WGBOOTSTRAP")
+	logCxt := log.WithFields(log.Fields{
+		"iface":    wgDeviceName,
+		"nodeName": nodeName,
+	})
 
-	if !configParams.WireguardHostEncryptionEnabled && !dbgBootstrap {
-		return nil
+	if !configParams.WireguardHostEncryptionEnabled {
+		// HostEncryption is currently enabled in environments by operator rather than through FelixConfiguration.
+		// This should not change for a given deployment. We only need to handle bootstrapping for clusters where
+		// host encryption is enabled.
+		logCxt.Debug("Host encryption is not enabled - no wireguard bootstrapping required")
+		return nil, nil
 	}
 
-	logCtx := log.WithFields(log.Fields{
-		"iface":    wgDeviceName,
-		"hostName": nodeName,
-		"ref":      "wgBootstrap",
+	logCxt.Debug("Bootstrapping wireguard")
+
+	if !configParams.WireguardEnabled || configParams.WireguardInterfaceName == "" {
+		logCxt.Info("Wireguard is not enabled - ensure no wireguard config")
+		return nil, RemoveWireguardForHostEncryptionBootstrapping(configParams, getNetlinkHandle, calicoClient)
+	}
+
+	// Get the local public key and the peer public keys currently programmed in the kernel.
+	kernelPublicKey, kernelPeerKeys := getWireguardInfo(logCxt, wgDeviceName, getWireguardHandle)
+
+	// If there is no valid wireguard configuration in the kernel then remove all traces of wireguard.
+	if kernelPublicKey == "" || kernelPeerKeys.Len() == 0 {
+		logCxt.Info("No valid wireguard kernel routing - removing wireguard configuration completely")
+		return nil, RemoveWireguardForHostEncryptionBootstrapping(configParams, getNetlinkHandle, calicoClient)
+	}
+
+	storedPublicKey, err := getPublicKeyForNode(logCxt, nodeName, calicoClient)
+	if err != nil {
+		// Could not determine the public key for this node.
+		return nil, fmt.Errorf("couldn't determine current wireguard configuration after %d retries: %v", bootstrapMaxRetries, err)
+	}
+
+	if storedPublicKey != kernelPublicKey {
+		// The public key configured in the kernel differs from the value stored in the node. Remove all wireguard
+		// configuration.
+		logCxt.Info("Found mismatch between kernel and datastore wireguard keys - removing wireguard configuration")
+		return nil, RemoveWireguardForHostEncryptionBootstrapping(configParams, getNetlinkHandle, calicoClient)
+	}
+
+	// The configured and stored wireguard key match.
+	logCxt.Infof("Wireguard public key matches kernel - we have %d configured peers", kernelPeerKeys.Len())
+	return kernelPeerKeys, nil
+}
+
+// FilterTyphaEndpoints filters the supplied set of typha endpoints to the set where wireguard routing is most likely
+// to succeed.
+func FilterTyphaEndpoints(
```

Comment:
I'm tweaking this so it doesn't return an error.  It's better to just include a typha endpoint than not include it when we hit an error.